lib/ddtrace/contrib/rails/action_view.rb

```diff
@@ -34,17 +34,39 @@ def self.get_key(f)
           'datadog_actionview_' + f
         end
 
-        def self.start_render_template(*)
-          key = get_key('render_template')
-          return if Thread.current[key]
+        def self.start_render_template(_name, _start, _finish, _id, payload)
+          # retrieve the tracing context
+          tracing_context = payload.fetch(:tracing_context)
 
+          # create a new Span and add it to the tracing context
           tracer = ::Rails.configuration.datadog_trace.fetch(:tracer)
-          type = Datadog::Ext::HTTP::TEMPLATE
-          tracer.trace('rails.render_template', span_type: type)
+          span = tracer.trace('rails.render_template', span_type: Datadog::Ext::HTTP::TEMPLATE)
+          tracing_context[:dd_rails_template_span] = span
+        rescue StandardError => e
+          Datadog::Tracer.log.debug(e.message)
+        end
 
-          Thread.current[key] = true
+        def self.finish_render_template(_name, _start, _finish, _id, payload)
+          # retrieve the tracing context and the latest active span
+          tracing_context = payload.fetch(:tracing_context)
+          span = tracing_context[:dd_rails_template_span]
+          return unless span && !span.finished?
+
+          # finish the tracing and update the execution time
+          begin
+            tracer = ::Rails.configuration.datadog_trace.fetch(:tracer)
+            template_name = tracing_context[:template_name]
+            layout = tracing_context[:layout]
+            exception = tracing_context[:exception]
+
+            span.set_tag('rails.template_name', template_name) if template_name
+            span.set_tag('rails.layout', layout) if layout
+            span.set_error(exception) if exception
+          ensure
+            span.finish()
+          end
         rescue StandardError => e
-          Datadog::Tracer.log.error(e.message)
+          Datadog::Tracer.log.debug(e.message)
```

Comment:
Are you sure you want a debug here? I'd assume debug is for messages that would otherwise clutter the log file with useless info, but this one should pretty much *never* be triggered. So needing to activate debug to see it might waste time (to go forth and back saying "you should activate debug mode") and also, the debug message might also appear with many other messages unrelated to this issue. All in all, I'd leave in debug, but there might be changes I'm not aware of.

Reply:
yes my idea was to keep `debug` all the time in our instrumentation, so that for our internal issues we're not polluting their logs (this is a hot-path so we may have thousands of line in the logs if we publish a wrong trace). On the other hand, I agree with you that it should *never* be triggered, so keeping an error could be useful. Probably I'll revert that to `log.error`.